content/advent-2017/go-execution-tracer.md

```diff
@@ -0,0 +1,237 @@
+
+Before gathering the trace, let's start by generating some load on our service using `wrk`: 
+
+```
+$ wrk -c 100 -t 10 -d 60s http://localhost:8181/hello
+```
+
+This will use 100 connections across 10 threads to make requests during 60 seconds. While `wrk` is running, we can collect 5s of trace data using `curl localhost:8181/debug/pprof/trace?seconds=5 > trace.out`. This generated a 5MB file (this can quickly grow if we are able to generate more load) on my 4 CPU machine.
+
+Once again, opening the trace is done by the go tool trace command: `go tool trace trace.out`. As the tool parses the entire content of the file, this will take longer than our previous example. When it completes, the page looks slightly different:
+
+```
+View trace (0s-2.546634537s)
+View trace (2.546634537s-5.00392737s)
+
+Goroutine analysis
+Network blocking profile
+Synchronization blocking profile
+Syscall blocking profile
+Scheduler latency profile
+```
+
+To guarantee that the browser will be able to render everything, the tool has divided the trace into two continuous parts. Busier applications or longer traces may require the tool to split this in even more parts. 
+
+Clicking on "View trace (2.546634537s-5.00392737s)" we can see that there is a lot going on:
+
+![View trace web](/postimages/advent-2017/go-execution-tracer/trace-web.png)
+
+This particular screenshot shows a GC run that starts at between 1169ms and 1170ms and ends right after 1174ms. During this time, an OS thread (PROC 1) ran a goroutine dedicated to the GC while other goroutines assisted in some GC phases (these are displayed on lines bellow the goroutine and are read MARK ASSIST). By the end of the screenshot, we can see that most of the allocated memory was freed by the GC.
+
+Another particular useful information is the number of goroutines that are in "Runnable" state (13 on the selected time): if this number becomes large over time this can indicate that we need more CPUs to handle the load.
+
+# Conclusions
+
+The tracer is a powerful tool for debugging concurrency issues, e.g, contentions and logical races. But it does not solves all problems: it is not the best tool available to track down what piece of code is spending most CPU time or allocations. The `go tool pprof` is better suited for these use cases.
+
+The tool really shines when you want to understand the behavior of a program over time and to know what each goroutine is doing when NOT running. Collecting traces may have some overhead and can generate a high amount of data to be inspected.
+
+Unfortunately, there is not many documentation available so some experimentation is needed to try and understand what the tracer is showing. This is also an [opportunity for contributions](https://github.com/golang/go/issues/16526) to the official documentation and to the community in general (e.g blog posts).
```

Comment:
I don't know a better way to say it, but "there is not many documentation available" doesn't seem right. Maybe "documentation is lacking" or so.